server/join.go

```diff
@@ -29,36 +30,118 @@ import (
 	endpoints := strings.Split(cfg.Join, ",")
 	return clientv3.Config{
 		Endpoints:   endpoints,
 		DialTimeout: defaultDialTimeout,
 	}
 }
 
-// prepareJoinCluster send MemberAdd command to pd cluster,
-// returns pd initial cluster configuration.
-func (cfg *Config) prepareJoinCluster() (string, error) {
+func memberAdd(client *clientv3.Client, urls []string) (*clientv3.MemberAddResponse, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), defaultDialTimeout)
 	defer cancel()
 
-	client, err := clientv3.New(cfg.genClientV3Config())
+	return client.MemberAdd(ctx, urls)
+}
+
+func memberList(client *clientv3.Client) (*clientv3.MemberListResponse, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultDialTimeout)
+	defer cancel()
+
+	return client.MemberList(ctx)
+}
+
+// prepareJoinCluster send MemberAdd command to pd cluster,
+// returns pd initial cluster configuration.
+//
+// TL;TR: The join functionality is safe. With data, join does nothing, w/o data
+//        and it is not a member of cluster, join does MemberAdd, otherwise
+//        return an error.
+//
+// Etcd automatically re-join cluster if there is a data dir, so first we check
+// if there is data dir or not. With data dir, it returns an empty string(etcd
+// will get correct configurations from data dir.)
+//
+// Without data dir, we may have following cases:
+//
+//  1. a new pd joins to an existing cluster.
+//      join does: MemberAdd, MemberList, then generate initial-cluster.
+//
+//  2. a new pd joins itself
+//      join does: nothing.
+//
+//  3. an failed pd re-joins to previous cluster.
+//      join does: return an error(etcd reports: raft log corrupted, truncated,
+//                 or lost?)
+//
+//  4. a join self pd failed and it restarted with join while other peers try
+//     to connect to it.
+//      join does: nothing. (join can not detect whether it is in a cluster or
+//                 not, however, etcd will handle it safey, if there is no data
+//                 in cluster the restarted pd will join to cluster, otherwise
+//                 pd will shutdown as soon as other peers connect to it. etcd
+//                 reports: raft log corrupted, truncated, or lost?)
+//
+//  5. a deleted pd joins to previous cluster.
+//      join does: same as case1. (it is not in member list and there is no
+//                 data, so we can treat it as a new pd.)
+//
+// With data dir, special case:
+//
+//  6. a failed pd tries to join to previous cluster but it has been deleted
+//     during it's downtime.
+//      join does: return "" (etcd will connect to other peers and will find
+//                 itself has been removed)
+//
+//  7. a deleted pd joins to previous cluster
+//      join does: return "" (as etcd will read data dir and find itself has
+//                 been removed, so an empty string is fine.)
+func (cfg *Config) prepareJoinCluster() (string, string, error) {
```

Comment:
I think we don't need to care some many cases.

1. if data-dir exists, let etcd handle that;
2. else if it join itself, create a new cluster for it;
3. else connect to the joined cluster and do `MemberAdd`, if it cannot connect to the cluster or the peer urls already exist (etcd will detect that), just let it failed.